src/worker/mod.rs

```diff
@@ -369,51 +412,86 @@ impl<
                 .await?;
         }
 
-        let exit = self.listen_for_heartbeats(&statuses).await;
+        // Either we have already taken this future and are polling it elsewhere (see Worker::run_to_completion),
+        // or they've never provided it to us (reminder: they do so via WorkerBuilder::with_graceful_shutdown)
+        let cancel_signal = self.shutdown_signal.take();
+
+        let report = tokio::select! {
+            // A signal from the user space received.
+            _ = async { let signal = cancel_signal.unwrap(); signal.await }, if cancel_signal.is_some() => {
+                let nrunning = tokio::select! {
+                     _ = tokio_sleep(self.shutdown_timeout.unwrap()), if self.shutdown_timeout.is_some() => {
+                        0
```

Comment:
A future TODO: it would be nice if we could give an estimate of how many workers actually remain here, but I think this is fine for now.

Reply:
added to the issues (#76)